tokio-net/src/driver/sharded_slab/mod.rs

```diff
@@ -0,0 +1,317 @@
+//! A lock-free concurrent slab.
+mod page;
+mod tid;
+pub(crate) use tid::Tid;
+mod iter;
+
+use crate::sync::{
+    atomic::{AtomicUsize, Ordering},
+    CausalCell,
+};
+use page::{slot, Addr, Page};
+use std::fmt;
+
+#[cfg(target_pointer_width = "64")]
+const MAX_THREADS: usize = 4096;
+#[cfg(target_pointer_width = "32")]
+const MAX_THREADS: usize = 2048;
+
+const MAX_PAGES: usize = WIDTH / 4;
+const RESERVED_BITS: usize = 5;
+
+const WIDTH: usize = std::mem::size_of::<usize>() * 8;
+
+/// A sharded, lock-free slab.
+pub(crate) struct Slab<T> {
+    shards: Box<[CausalCell<Shard<T>>]>,
+}
+
+struct Shard<T> {
+    #[cfg(debug_assertions)]
+    tid: usize,
+    sz: usize,
+    len: AtomicUsize,
+    // ┌─────────────┐      ┌────────┐
+    // │ page 1      │      │        │
+    // ├─────────────┤ ┌───▶│  next──┼─┐
+    // │ page 2      │ │    ├────────┤ │
+    // │             │ │    │XXXXXXXX│ │
+    // │ local_free──┼─┘    ├────────┤ │
+    // │ global_free─┼─┐    │        │◀┘
+    // ├─────────────┤ └───▶│  next──┼─┐
+    // │   page 3    │      ├────────┤ │
+    // └─────────────┘      │XXXXXXXX│ │
+    //       ...            ├────────┤ │
+    // ┌─────────────┐      │XXXXXXXX│ │
+    // │ page n      │      ├────────┤ │
+    // └─────────────┘      │        │◀┘
+    //                      │  next──┼───▶
+    //                      ├────────┤
+    //                      │XXXXXXXX│
+    //                      └────────┘
+    //                         ...
+    pages: Vec<Page<T>>,
+}
+
+impl<T> Slab<T> {
+    pub(crate) const MAX_BIT: usize = slot::Generation::LEN + slot::Generation::SHIFT;
+
+    /// Returns a new slab.
+    pub(crate) fn new() -> Self {
+        let mut shards = Vec::with_capacity(MAX_THREADS);
+        let mut idx = 0;
+        shards.resize_with(MAX_THREADS, || {
+            let shard = Shard::new(idx);
+            idx += 1;
+            CausalCell::new(shard)
+        });
+        Slab {
+            shards: shards.into_boxed_slice(),
+        }
+    }
+
+    /// Inserts a value into the slab, returning a key that can be used to
+    /// access it.
+    ///
+    /// If this function returns `None`, then the shard for the current thread
+    /// is full and no items can be added until some are removed, or the maximum
+    /// number of shards has been reached.
+    pub(crate) fn insert(&self, value: T) -> Option<usize> {
+        let tid = Tid::current();
```

Comment:
I think we should aim to get rid of this. I don't know if it should be this PR or a follow up PR (after some other work has landed).

Ideally, we would be able to avoid thread-locals at this point. We usually hit a thread-local to get to this point. However, to do this probably would require more restructuring of runtime constructs, so we might just want to punt.